Fix(quadlet): Ensure .app application file is removed by quadlet rm --all (#27647)

kavishgr opened the pull request:
- [x] Certify you wrote the patch or otherwise have the right to pass it on as an open-source patch by signing all
commits. (`git commit -s`). (If needed, use `git commit -s --amend`).  The author email must match
the sign-off email address. See [CONTRIBUTING.md](https://github.com/containers/podman/blob/main/CONTRIBUTING.md#sign-your-prs)
for more information.
- [x] Referenced issues using `Fixes: #00000` in commit message (if applicable)
- [x] [Tests](https://github.com/containers/podman/tree/main/test#readme) have been added/updated (or no tests are needed)
- [x] [Documentation](https://github.com/containers/podman/blob/main/docs/README.md) has been updated (or no documentation changes are needed)
- [x] All commits pass `make validatepr` (format/lint checks)
- [x] [Release note](https://github.com/kubernetes/community/blob/master/contributors/guide/release-notes.md) entered in the section below (or `None` if no user-facing changes)

#### Does this PR introduce a user-facing change?

```release-note
None
```


Fixes: #27647

**Note on Dependencies**:

This PR is dependent on the changes from issue #27653, as outlined in [PR](https://github.com/containers/podman/pull/27693) #27693. Once #27693 merges, I will rebase and push the changes here.


What I did (`quadlet.go` file):

- Fixed `reverseMap` lookup by using the base filename instead of the full path, since r`everseMap` keys are filenames only

- Added conditional path handling when expanding quadlets from .app files - use full paths when --all flag is set (since `getAllQuadletPaths` returns full paths), but use just filenames otherwise (since `getQuadletPathByName` expects filenames)

- Moved the cleanup of .app and .asset metadata files outside the main quadlet removal loop. Previously these files were being left behind because they aren't included in `allQuadletPaths`. Now they're properly removed after all quadlets have been processed

- Added logging for the files being removed

These changes fix the issue where .`app` files were not being deleted during removal(`podman quadlet --all --force`), and also resolve path resolution errors when removing individual apps (e.g., podman quadlet rm .nginx-test.app).

Test: 

- The test verifies that the `podman quadlet rm -af` command correctly removes all quadlet files.

kavishgr committed:
Fix(quadlet): Deduplicate removal files for .app expansion (#27653)

Refactor removelist into sets

Add quadlet rm --all test with safename

add an integration test for quadlet rm --all(skip for remote clients)

Signed-off-by: Kavish Gour <kavishgr@protonmail.com>

Honny1 reviewed (commented):
I did a quick check of the CI failure. I also fixed the release note format in the PR description.

Honny1 reviewed (commented):
The code LGTM. However, I have a question about the tests: is there a specific reason they weren't implemented in `quadlet_test.go`?

ygalblum reviewed (commented):
Small nit

Honny1 reviewed (approved):
LGTM, please rebase on main.

Honny1 commented on test/e2e/quadlet_test.go:
```diff
@@ -1334,4 +1334,51 @@ BOGUS=foo
+			installSession := podmanTest.Podman([]string{"quadlet", "install", tmpDir})
+			installSession.WaitWithDefaultTimeout()
+			Expect(installSession).Should(ExitCleanly())
+
```
You can simplify this by using `podmanTest.PodmanExitCleanly`.

Honny1 commented:
Not sure about the F43 failure, but the other tests appear to be flaky.

```
[FAIL] quadlet system generator Running quadlet force remove all [It] Should remove all quadlets at once
[+0894s]   /var/tmp/go/src/github.com/containers/podman/test/e2e/quadlet_test.go:1360
```

PTAL @mheon @Luap99

kavishgr commented:
Ping. Is there anything else you need from my side to move this forward ?

Honny1 reviewed (commented):
LGTM. Just a quick pass from me. 

PTAL @containers/podman-maintainers

giuseppe reviewed (approved):
LGTM

kavishgr commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -777,32 +781,36 @@ func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string,
-	quadlets = expandQuadletList
+
+	// Convert expandQuadletSet to slice
+	quadlets = make([]string, 0, len(expandQuadletSet))
```
You're right, thanks for catching that. 
I've moved the check before the conversion. It's fixed now.

Honny1 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -891,30 +903,37 @@ func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string,
+				// this ensures consistency since getAllQuadletPaths() returns full paths
+				// but getQuadletPathByName() expects filenames
+				var entryToAdd string
+				if options.All {
```
When the initial `.app` expansion loop runs, it adds `filenames` (e.g. `foo.container`) into `expandQuadletSet`. Then this `options.All` block adds full paths (e.g. `/etc/containers/systemd/foo.container`) from `getAllQuadletPaths()`. The dedup check won't detect these refer to the same file, so a quadlet could be processed twice.

In practice, `--all` is unlikely to be combined with explicit `.app` arguments from the CLI, so this is low-risk, but it would be worth either:
- Normalizing all entries to full paths when `options.All` is `true`, or
- Adding a comment documenting that `--all` and explicit arguments are mutually exclusive

Honny1 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -990,13 +1009,24 @@ func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string,
 
+	// Remove .app and .asset files after the main quadlet removal loop
+	// This ensures they are cleaned up properly since they are not included in allQuadletPaths
+	installDir := systemdquadlet.GetInstallUnitDirPath(rootless.IsRootless())
```
This expression appears several times throughout `QuadletRemove`. You already computed it once here as `installDir`. I would put that at the top of the function and reuse it everywhere.

kavishgr committed:
quadlet-rm: disallow combining --all flag with individual quadlets

Ensure that the --all (-a) flag cannot be used simultaneously with specific quadlet arguments on the CLI. This prevents logical conflicts when a user provides both the --all flag and specific quadlets.

- Add validation check in cmd/podman/quadlet/remove.go
- Update the docs to document this behaviour with an example
- Improve user feedback by explicitly stating why the command failed

Signed-off-by: Kavish Gour <kavishgr@protonmail.com>

kavishgr commented:
Sorry about that. Ready for another look.

Honny1 commented:
PTAL @containers/podman-maintainers

Honny1 commented on test/e2e/quadlet_test.go:
```diff
@@ -1371,4 +1373,49 @@ BOGUS=foo
 		),
 	)
+	Describe("Running quadlet force remove all", func() {
+		It("Should remove all quadlets at once", func() {
+			SkipIfRemote("quadlet is not supported for remote clients")
+			SkipIfSystemdNotRunning("quadlet install requires a running systemd")
```
```suggestion
			SkipIfRemote("quadlet is not supported for remote clients")
			SkipIfInContainer("quadlet install requires a running systemd")
			SkipIfSystemdNotRunning("quadlet install requires a running systemd")
```

kavishgr replied:
Yes, I guess this is it.

kavishgr replied:
Makes total sense. I guess this is it. Thanks.

Honny1 reviewed (changes requested):
I think I found it.

Honny1 reviewed (approved):
Thanks, LGTM

PTAL @containers/podman-maintainers

Honny1 commented on cmd/podman/quadlet/remove.go:
```diff
@@ -49,6 +49,9 @@ func rm(_ *cobra.Command, args []string) error {
 		return errors.New("at least one quadlet file must be selected")
 	}
+	if len(args) > 0 && removeOptions.All {
+		return errors.New("-a or --all cannot be used when combined with individual quadlets")
```
Non-blocking:
```suggestion
		return errors.New("--all (-a) cannot be combined with individual quadlet arguments")
```

Honny1 commented on cmd/podman/quadlet/remove.go:
```diff
@@ -49,6 +49,9 @@ func rm(_ *cobra.Command, args []string) error {
 	if len(args) < 1 && !removeOptions.All {
 		return errors.New("at least one quadlet file must be selected")
 	}
+	if len(args) > 0 && removeOptions.All {
```
I would add the same check for the REST api.

kavishgr replied:
I guess we already have this check in [pkg/api/handlers/libpod/quadlets.go#L343-L347](https://github.com/containers/podman/blob/main/pkg/api/handlers/libpod/quadlets.go#L343-L347):

```go
	// Validate that both all and quadlets are not provided together
	if query.All && len(query.Quadlets) > 0 {
		utils.Error(w, http.StatusBadRequest, fmt.Errorf("when setting all=true, you may not pass any quadlet names"))
		return
	}
```

Honny1 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -913,30 +927,43 @@ func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string,
-				if !slices.Contains(quadlets, entry) {
-					quadlets = append(quadlets, entry)
+				var entryToAdd string
+				// Note: We treat --all and specific arguments (e.g. foo.container)
```
Nit: This comment describes a scenario where --all is combined with explicit .app arguments, but that combination is now rejected at the CLI layer in remove.go. This makes most of the comment stale.

Honny1 commented on pkg/domain/infra/abi/quadlet.go:
```diff
@@ -1012,13 +1039,23 @@ func (ic *ContainerEngine) QuadletRemove(ctx context.Context, quadlets []string,
 
+	// Remove .app and .asset files after the main quadlet removal loop
+	// This ensures they are cleaned up properly since they are not included in allQuadletPaths
+	for entry := range removeSet {
```
Not blocking: metadata files in removeSet are cleaned up regardless of whether the associated quadlet removals succeeded. If all quadlets in an app fail to remove (e.g., running without --force), the .app file is still deleted, orphaning them. This was the same behavior before this PR, just noting it as something to potentially address in a follow-up.
